WordPress/gutenberg
# Properly apply styles and classes to the experimental form block

## What?
Properly applies styles and classes to the experimental form block. In https://github.com/WordPress/gutenberg/pull/44214#issuecomment-1786948891 it shows that styles don't get properly applied to the `core/form` block. This PR fixes that issue.

## Why?
Because it's a bug - or rather something that wasn't properly implemented in the initial PR introducing the experiment.

## How?
* ~Removes the default `style` and `classes` attributes from the `<form>` element, and then adds them using `get_block_wrapper_attributes()`.~

---

Updated by @t-hamano:

- Remove the hard-coded `className` prop in the save function. This means that styles, classes, etc. from useblockProps will be respected, and block support will work properly.
- Add a block deprecation process so as not to break existing blocks in the contents. This deprecation is unusual and intentionally leaves the block support empty. See [this comment](https://github.com/WordPress/gutenberg/pull/55755/files#diff-f5e02f945f8e2ade1188bf021549961746ef35a7b7d1ae6e0c7a43d3b7dce201R13-R23) for the reason.

---

## Testing Instructions
* Enable the forms-blocks experiment, create a post and add a form in there.
* Change the background colors, padding, margin, typography... whatever you want.
* Save the post and visit it on the front (in the editor it was working properly, but not when viewing the post)
* Confirm that your styles were properly applied.

---

Added by @t-hamano:

### Testing for Block Deprecation

- Check out the trunk branch.
- Insert a Form block in the Post Editor. Apply a background color or change the font size.
- Save the post.
- Check out this branch.
- Reload the browser in the Post Editor.
- The block should not break, and your browser console logs that the migration process has been done correctly.
---

### Testing Instructions for Keyboard
Not applicable
